_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,281 @@
+
+And what happens when offsets are no longer valid? If a consumer group or standalone consumer is inactive and commits no offsets during the _offsets retention period_ (`offsets.retention.minutes`) configured for a broker, previously committed offsets are deleted from `__consumer_offsets`. Again, you can use the `earliest` option in this situation so that the consumer returns to the start of a partition to avoid data loss if offsets were not committed.
+
+If the amount of data returned in a single fetch request is large, a timeout might occur before the consumer has processed it. In which case, you can lower `max.partition.fetch.bytes` or increase `session.timeout.ms` as part of you offset policy.
+
+```properties
+# ...
+auto.offset.reset=earliest
+session.timeout.ms=10000
+max.partition.fetch.bytes=1048576
+# ...
+```
+
+## Minimizing the impact of rebalancing your consumer group
+
+Rebalancing is the time taken to assign a partition to active consumers in a group. During a rebalance:
+
+* Consumers commit their offsets
+* A new consumer group is formed
+* The designated group leader assigns partitions evenly to group members
+* The consumers in the group receive their assignments and start to fetch data
+
+Obviously, the rebalancing process takes time. And during a rolling restart of a consumer group cluster, it happens repeatedly. So rebalancing can have a clear impact on the performance of your cluster group. As mentioned, increasing the number of heartbeat checks reduces the likelihood of unnecessary rebalances. But there a couple more approaches you can take. One is to set up _static membership_ to reduce the overall number of rebalances. Static membership uses persistence so that a consumer instance is recognized during a restart after a session timeout.
+
+You can use the `group.instance.id` property to specify a unique group instance id for a consumer. The consumer group coordinator can then use the id when identifying a new consumer instance following a restart.
+
+```properties
+# ...
+group.instance.id=UNIQUE-ID
+# ...
+```
+
+The consumer group coordinator assigns the consumer instance a new member id, but as a static member it continues with the same instance id, and receives the same assignment of topic partitions is made.
+
+You can also calibrate the `max.poll.interval.ms` and `max.poll.interval.ms` polling configuration properties to reduce the number of rebalances.
```

Comment:
Changing these only really helps when the cause of the rebalances is due to the poll interval. IOW, it's only indicated when the logs demontrated that this is the problem.

Reply:
Changed to:

> Another cause of rebalancing might actually be due to an insufficient poll interval configuration, which is then interpreted as a consumer failure. If you discover this situation when examining your consumer logs, you can calibrate the `max.poll.interval.ms` and `max.poll.interval.ms` polling configuration properties to reduce the number of rebalances.